modules/internal/ChapelLocale-chpldoc.chpl

```diff
@@ -0,0 +1,608 @@
+      // so tell the compiler to not insert them.
+      pragma "no copy" pragma "no auto destroy"
+      const origLocales => (origRootLocale:RootLocale).getDefaultLocaleArray();
+      var origRL = origLocales._value.theData;
+      var newRL = newRootLocale.getDefaultLocaleArray()._value.theData;
+      // We must directly implement a bulk copy here, as the mechanisms
+      // for doing so via a whole array assignment are not initialized
+      // yet and copying element-by-element via a for loop is is costly.
+      __primitive("chpl_comm_get",
+                  __primitive("array_get", newRL, 0),
+                  0 /* locale 0 */,
+                  __primitive("array_get", origRL, 0), 
+                  numLocales.safeCast(size_t));
+      // Set the rootLocale to the local copy
+      rootLocale = newRootLocale;
+    }
+    if locIdx!=0 {
+      // We mimic a private Locales array alias by using the move
+      // primitive.
+      __primitive("move", Locales,
+                  (rootLocale:RootLocale).getDefaultLocaleArray());
+    }
+  }
+
+  // We need a temporary value for "here" before the architecture is defined.
+  // This is due to the fact that "here" is used for memory and task control
+  // in setting up the architecture itself.
+  // Its type should probably be renamed dummyLocale or something
+  // representative.
+  // The dummy locale provides system-default tasking and memory management.
+  pragma "no doc"
+  const dummyLocale = new locale();
+
+  pragma "no doc"
+  extern proc chpl_task_getRequestedSubloc(): chpl_sublocID_t;
+
+// Commented out because causes chpldoc error message:
+//   The name ... cannot be exported twice from the same compilation unit.
+//
+//  pragma "no doc"
+//  pragma "insert line file info"
+//  export
+//  proc chpl_getLocaleID(ref localeID: chpl_localeID_t) {
+//    localeID = here_id;
+//  }
+
+  // Return the locale ID of the current locale
+  pragma "no doc"
+  inline proc here_id {
+     if localeModelHasSublocales then
+      return chpl_rt_buildLocaleID(chpl_nodeID, chpl_task_getRequestedSubloc());
+    else
+      return chpl_rt_buildLocaleID(chpl_nodeID, c_sublocid_any);
+  }
+
+  /*
+    This returns the locale from which the call is made.
+
+    :return: current locale
+    :rtype: locale
+  */
+  inline proc here {
+    return chpl_localeID_to_locale(here_id);
+  }
+  
+  // Returns a wide pointer to the locale with the given id.
+  pragma "no doc"
+  proc chpl_localeID_to_locale(id : chpl_localeID_t) : locale {
+    if rootLocale then
+      return (rootLocale:AbstractRootLocale).localeIDtoLocale(id);
+    else
+      // For code prior to rootLocale initialization
+      return dummyLocale;
+  }
+
+  /*
+    ``numLocales`` is the number of top-level (network connected)
+    locales.
+  */  
+  config const numLocales: int;
+
+  /*
+    This is the domain over which the global ``Locales`` array is
+    defined.
+  */  
+  const LocaleSpace: domain(1) = [0..numLocales-1];
```

Comment:
Hmm this renders really ugly because at parse time we replace the array syntax with the builder routines.. 

Also, I know it's in the spec as you wrote it, but shouldn't this be:

 ``` chapel
const LocaleSpace = {0..numLocales-1};
```

I think we can do some post-processing to clean up how it gets generated. Let me take a look at that